pkg/auth/providers/azure/azure_provider.go

```diff
@@ -211,154 +217,154 @@ func (ap *azureProvider) TransformToAuthProvider(
 	}
-
-	return userPrincipal, groupPrincipals, string(providerToken), nil
 }
 
-func (ap *azureProvider) getUser(client *azureClient, principalID string, token v3.Token) (v3.Principal, error) {
-	user, err := client.userClient.Get(context.Background(), principalID)
-	if err != nil {
-		return v3.Principal{}, httperror.NewAPIError(httperror.NotFound, err.Error())
+func (ap *azureProvider) msGraphUsersToPrincipals(users *[]msgraph.User, token v3.Token) ([]v3.Principal, error) {
+	var principals []v3.Principal
+	for _, u := range *users {
+		p, err := userToPrincipal(&u)
+		if err != nil {
+			return nil, err
+		}
+		p.Me = samePrincipal(&token.UserPrincipal, p)
+		principals = append(principals, *p)
 	}
-
-	p := ap.userToPrincipal(user)
-	p.Me = samePrincipal(token.UserPrincipal, p)
-
-	return p, nil
+	return principals, nil
 }
 
-func (ap *azureProvider) getGroup(client *azureClient, principalID string, token v3.Token) (v3.Principal, error) {
-	group, err := client.groupClient.Get(context.Background(), principalID)
-	if err != nil {
-		return v3.Principal{}, httperror.NewAPIError(httperror.NotFound, err.Error())
+func (ap *azureProvider) azureADGraphUsersToPrincipals(users graphrbac.UserListResultPage, token v3.Token) ([]v3.Principal, error) {
+	var principals []v3.Principal
+	for _, u := range users.Values() {
+		p, err := userToPrincipal(u)
+		if err != nil {
+			return nil, err
+		}
+		p.Me = samePrincipal(&token.UserPrincipal, p)
+		principals = append(principals, *p)
 	}
-
-	p := ap.groupToPrincipal(group)
-	p.MemberOf = ap.tokenMGR.IsMemberOf(token, p)
-
-	return p, nil
+	return principals, nil
 }
 
-func (ap *azureProvider) searchUsers(client *azureClient, name string, token v3.Token) ([]v3.Principal, error) {
-	filter := fmt.Sprintf("startswith(userPrincipalName,'%[1]s') or startswith(displayName,'%[1]s') or startswith(givenName,'%[1]s') or startswith(surname,'%[1]s')", name)
-	users, err := client.userClient.List(context.Background(), filter, "")
-	if err != nil {
-		return nil, err
+func (ap *azureProvider) msGraphGroupsToPrincipals(groups *[]msgraph.Group, token v3.Token) ([]v3.Principal, error) {
+	var principals []v3.Principal
+	for _, group := range *groups {
+		p, err := groupToPrincipal(&group)
+		if err != nil {
+			return nil, err
+		}
+		p.MemberOf = ap.tokenMGR.IsMemberOf(token, *p)
+		principals = append(principals, *p)
 	}
+	return principals, nil
+}
 
+func (ap *azureProvider) azureADGraphGroupsToPrincipals(groups graphrbac.GroupListResultPage, token v3.Token) ([]v3.Principal, error) {
 	var principals []v3.Principal
-	for _, user := range users.Values() {
-		p := ap.userToPrincipal(user)
-		p.Me = samePrincipal(token.UserPrincipal, p)
-		principals = append(principals, p)
+	for _, group := range groups.Values() {
+		p, err := groupToPrincipal(group)
+		if err != nil {
+			return nil, err
+		}
+		p.MemberOf = ap.tokenMGR.IsMemberOf(token, *p)
+		principals = append(principals, *p)
 	}
 	return principals, nil
 }
```

Comment:
By using interfaces and combining userToPrincipal and groupToPrincipal could you combine these into 1 or 2 functions? This would help code maintenance by making it so if a change is made, it is made for all types at once since any change would probably apply to all of them.